Install CMake export files

stefanosoffia opened the pull request:
_CMake export files_ automate the task of importing GTest libraries into the CMake project that links them (See [Exporting from an Installation Tree](https://cmake.org/Wiki/CMake/Tutorials/Exporting_and_Importing_Targets#Exporting_from_an_Installation_Tree)). Also:

1.  Issue https://github.com/google/googletest/issues/878 mentions two alternative approaches, but the _CMake export files_ are the more idiomatic solution nowadays (see [The Ultimate Guide to Modern CMake](https://rix0r.nl/blog/2015/08/13/cmake-guide/)). Specifically:
  a. The `add_subdirectory` approach might interfere with the global settings of the host CMake project.
  b. The `find_package` approach does not support Release+Debug installations.
  The _CMake export files_ address both these limitations.
2. W.r.t. PR #768, this patch uses 'GTest' instead of 'Gtest' for the CMake namespace of the exported targets. Rationale: to be consistent with the existing 'FindGtest' CMake module (https://cmake.org/cmake/help/v3.10/module/FindGTest.html).
3. Also, w.r.t. PR #768, this patch does not override the `DEBUG_POSTFIX` CMake property for the exported targets. Rationale: give to the user the flexibility to choose its own suffix (and if to use one at all).

stefanosoffia commented:
I signed it!

headupinclouds commented:
Any plans to merge this?

stefanosoffia commented:
I'm still available. Just rebased against master. Let me know if any clarification is needed.

sebastianlipponer commented on googletest/CMakeLists.txt:
```diff
@@ -131,6 +139,12 @@ if(INSTALL_GTEST)
+  install(EXPORT GtestTargets
+    FILE GtestConfig.cmake
+    NAMESPACE GTest::
+    DESTINATION cmake)
```
`DESTINATION` should take `${CMAKE_INSTALL_LIBDIR}` into account. Therefore, change this line to:
`DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/Gtest`

sebastianlipponer commented on googlemock/CMakeLists.txt:
```diff
@@ -129,8 +136,15 @@ if(INSTALL_GMOCK)
+  install(EXPORT GtestTargets
+    FILE GmockConfig.cmake
+    NAMESPACE GTest::
+    DESTINATION cmake)
```
`DESTINATION` should take `${CMAKE_INSTALL_LIBDIR}` into account. Otherwise its not consistent with the pkgconfig installation path and follows a path convention specfic to Windows systems (see [find_package](https://cmake.org/cmake/help/v3.8/command/find_package.html#command:find_package)). Therefore, change this line to:
`DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/Gtest`
which is used on Windows and Unix systems.

sebastianlipponer commented on googlemock/CMakeLists.txt:
```diff
@@ -103,15 +103,22 @@ cxx_library(gmock_main
 ########################################################################
 #
 # Install rules
 if(INSTALL_GMOCK)
```
I do miss the installation of a `GtestConfigVersion.cmake` file for version information. Please add:
```
  include(CMakePackageConfigHelpers)
  write_basic_package_version_file(
    "${CMAKE_CURRENT_BINARY_DIR}/GtestConfigVersion.cmake"
    COMPATIBILITY AnyNewerVersion)
```

sebastianlipponer commented on googlemock/CMakeLists.txt:
```diff
@@ -129,8 +136,15 @@ if(INSTALL_GMOCK)
   install(FILES "${CMAKE_BINARY_DIR}/gmock.pc" "${CMAKE_BINARY_DIR}/gmock_main.pc"
     DESTINATION "${CMAKE_INSTALL_LIBDIR}/pkgconfig")
+
+  # install CMake export files
```
```
  install(FILES "${CMAKE_CURRENT_BINARY_DIR}/GtestConfigVersion.cmake"
    DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/Gtest")
```

sebastianlipponer commented on googletest/CMakeLists.txt:
```diff
@@ -105,15 +105,23 @@ target_link_libraries(gtest_main gtest)
 ########################################################################
 #
 # Install rules
 if(INSTALL_GTEST)
```
I do miss the installation of a `GtestConfigVersion.cmake` file for version information. Please add:
```
  include(CMakePackageConfigHelpers)
  write_basic_package_version_file(
    "${CMAKE_CURRENT_BINARY_DIR}/GtestConfigVersion.cmake"
    COMPATIBILITY AnyNewerVersion)
```

sebastianlipponer commented on googletest/CMakeLists.txt:
```diff
@@ -131,6 +139,12 @@ if(INSTALL_GTEST)
   install(FILES "${CMAKE_BINARY_DIR}/gtest.pc" "${CMAKE_BINARY_DIR}/gtest_main.pc"
     DESTINATION "${CMAKE_INSTALL_LIBDIR}/pkgconfig")
+
+  # install CMake export files
```
```
  install(FILES "${CMAKE_CURRENT_BINARY_DIR}/GmockConfigVersion.cmake"
    DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/Gmock")
```

sebastianlipponer commented:
As already mentioned in #768 you should consider using `configure_package_config_file()` in order generate the `GtestConfig.cmake` file (see [CMakePackageConfigHelpers](https://cmake.org/cmake/help/v3.8/module/CMakePackageConfigHelpers.html). This file will then be relocateable.

stefanosoffia commented:
Thanks @sebastianlipponer for all your feedbacks. I updated the PR accordingly: I added the `ConfigVersion.cmake` generation, renamed and moved the `Config.cmake` files as you described, so as to enable `find_package(CONFIG)`. I also moved the code to a shared function used for both `Ctest` and `Cmock`.

Please, help me understand the last detail. W.r.t. the export files created by the `install(EXPORT)`:
* By looking at their content, and also by reading from the [cmake packaging manual](https://cmake.org/cmake/help/latest/manual/cmake-packages.7.html#creating-relocatable-packages), it seems to me that these files are already _relocatable_. Specifically, see the use of `$<INSTALL_INTERFACE>` for exported targets, which enables the generated `install(EXPORT)` file to be relocatable.
* As you pointed out, a comment in the discussion of PR #768 references the [Hunter fork of Google Test](https://github.com/hunter-packages/googletest) as an example of use of [`configure_package_config_file`](https://cmake.org/cmake/help/latest/module/CMakePackageConfigHelpers.html#command:write_basic_package_version_file). Though, Hunter's implementation seems to just [wrap](https://github.com/hunter-packages/googletest/blob/hunter/googletest/cmake/Config.cmake.in) the generated `install(EXPORT) file, see [`add_install_rules`](https://github.com/hunter-packages/googletest/blob/hunter/googletest/cmake/internal_utils.cmake) function.

ruslo commented:
> but we do not have components in this project

It's a checker. When user do `find_package(GTest CONFIG REQUIRED my-mistake)` the `my-mistake` string will be interpreted as a component. And without checker it will be silently ignored.

> So, I am currently failing to see the contribution of configure_package_config_file() to this project. Notice that the cmake packaging manual does not even mention it

Here: https://cmake.org/cmake/help/latest/module/CMakePackageConfigHelpers.html#command:configure_package_config_file

If you want to touch a real code, take a look at this example:
* https://github.com/forexample/package-example

And this boilerplate part:
* https://github.com/forexample/package-example/blob/8b4b91bc6f3bc61de958e61536f63fd35a2648a3/Foo/CMakeLists.txt#L69-L153

ruslo commented on googlemock/CMakeLists.txt:
```diff
@@ -103,20 +103,28 @@ cxx_library(gmock_main
+      $<BUILD_INTERFACE:${gmock_include_dirs}>
+      $<INSTALL_INTERFACE:include>)
+  endif()
+  target_include_directories(gmock      SYSTEM INTERFACE ${gmock_include_dirs})
```
`SYSTEM` should be removed because `$<BUILD_INTERFACE:${gmock_include_dirs}>` will be used while building

ruslo replied:
I think `PUBLIC` is better suitable here, instead of `INTERFACE`.

stefanosoffia replied:
@ruslo, I think you are right here. However, both the occurrence of `SYSTEM` and the use of `INTERFACE` predates this PR. I agree that they should be fixed, but:
* Not sure why `SYSTEM` is there. On my setups it works also without it but removing it might introduce issues with other compilers, i.e. it shall be discussed separately.
* AFAIK, to properly use $<BUILD_INTERFACE> the `include_directories()` occurrences should be replaced with `target_include_directories()`. But this task is not trivial because it has side effects on the tests. Also, it would require either additional CMake version switches or to update the minimum required version to something more recent than `2.6.4`. Personally, I would be in favor of the second, though I also understand it might have undesired side-effects I might not be aware of. AFAIK, also this shall be discussed in a separate PR.
* They are both not essential to the goal of this PR. Therefore, in the current PR proposal I dropped the `$<BUILD_INTERFACE>` occurrences because they were redundant with the existing `include_directories()` and only left the `$<INSTALL_INTERFACE>` which is required to complete this PR.

If you agree I would like to defer further changes to subsequent PRs.

stefanosoffia commented on googlemock/CMakeLists.txt:
```diff
@@ -103,20 +103,28 @@ cxx_library(gmock_main
-  target_include_directories(gmock      SYSTEM INTERFACE "${gmock_SOURCE_DIR}/include")
-  target_include_directories(gmock_main SYSTEM INTERFACE "${gmock_SOURCE_DIR}/include")
+  set(gmock_include_dirs "${gmock_SOURCE_DIR}/include")
+  if (NOT "${CMAKE_VERSION}" VERSION_LESS "3.0.0")
```
Ack. We can safely remove the check at line 107.

stefanosoffia commented:
Thanks @ruslo for your feedbacks too. Now I better understand the role of `configure_package_config_file()` and specifically the use of `check_required_components()` to validate the `find_package` invocation. Therefore I added the separate generation of the `Config.cmake` and `Targets.cmake` files as documented in the linked `package-example`.

Please, also see the above replies to your other comments.

stefanosoffia commented on googlemock/CMakeLists.txt:
```diff
@@ -103,20 +103,28 @@ cxx_library(gmock_main
+      $<INSTALL_INTERFACE:include>)
+  endif()
+  target_include_directories(gmock      SYSTEM INTERFACE ${gmock_include_dirs})
+  target_include_directories(gmock_main SYSTEM INTERFACE ${gmock_include_dirs})
```
Same as above. It _should_ but AFAIK it does not in the current implementation. Can we move this to separate PR?

ruslo commented on googlemock/CMakeLists.txt:
```diff
@@ -103,20 +103,23 @@ cxx_library(gmock_main
 if (DEFINED CMAKE_VERSION AND NOT "${CMAKE_VERSION}" VERSION_LESS "2.8.11")
-  target_include_directories(gmock      SYSTEM INTERFACE "${gmock_SOURCE_DIR}/include")
-  target_include_directories(gmock_main SYSTEM INTERFACE "${gmock_SOURCE_DIR}/include")
+  set(gmock_include_dirs $<INSTALL_INTERFACE:include>)
```
If local directory "${gmock_SOURCE_DIR}/include" is not needed then all this block can be removed and `INCLUDES DESTINATION include` used instead in `install`, see:
* https://github.com/forexample/package-example/blob/8b4b91bc6f3bc61de958e61536f63fd35a2648a3/Foo/CMakeLists.txt#L111-L118

Will achieve the same but in cleaner way. `INCLUDES DESTINATION` available since CMake 2.8.12:
* https://cmake.org/cmake/help/v2.8.12/cmake.html#command:install

stefanosoffia replied:
Agreed. Updated accordingly.

ruslo commented on googletest/cmake/internal_utils.cmake:
```diff
@@ -271,3 +271,32 @@ function(py_test name)
+  # Wrap the exported targets in the CMake package utility.
+  set(config_file "${generated_dir}/${package_name}Config.cmake")
+  configure_package_config_file(
+    "${CMAKE_SOURCE_DIR}/googletest/cmake/Config.cmake.in"
```
Variable `CMAKE_SOURCE_DIR` is a top-level directory of source tree. This code will not work correctly if GTest will be added by `add_subdirectory`. You can use `${gtest_SOURCE_DIR}/googletest/cmake/Config.cmake.in` instead.

stefanosoffia replied:
Agreed. Fixed.

sebastianlipponer commented:
@ruslo Great comments. Thank you for reviewing.

stefanosoffia commented:
@ruslo, thanks again. Agreed to add `PUBLIC` to the `target_link_libraries()`, but it is not clear to me what you mean with the _populate definitions_ comment.

stefanosoffia commented on googletest/cmake/Config.cmake.in:
```diff
@@ -0,0 +1,7 @@
+@PACKAGE_INIT@
+include(CMakeFindDependencyMacro)
+set(THREADS_PREFER_PTHREAD_FLAG @THREADS_PREFER_PTHREAD_FLAG@)
+find_dependency(Threads)
```
According to the [cmake packaging docs](https://cmake.org/cmake/help/latest/manual/cmake-packages.7.html#creating-a-package-configuration-file), since `gtest` links publicly `Threads::Threads`, then the exported `Config.cmake` file shall also call the `find_dependency(Threads)` so as to satisfy all gtest dependencies. Do you agree on this?

stefanosoffia replied:
Ack. Modified as suggested.

stefanosoffia commented on googletest/cmake/internal_utils.cmake:
```diff
@@ -161,7 +161,7 @@ function(cxx_library_with_type name type cxx_flags)
   endif()
   if (DEFINED GTEST_HAS_PTHREAD)
-    target_link_libraries(${name} ${CMAKE_THREAD_LIBS_INIT})
+    target_link_libraries(${name} PUBLIC Threads::Threads)
```
Ack. Added cmake version checking.

stefanosoffia commented on googlemock/CMakeLists.txt:
```diff
@@ -106,24 +106,19 @@ else()
     LIBRARY DESTINATION "${CMAKE_INSTALL_LIBDIR}"
-    ARCHIVE DESTINATION "${CMAKE_INSTALL_LIBDIR}")
+    ARCHIVE DESTINATION "${CMAKE_INSTALL_LIBDIR}"
+    INCLUDES DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}")
```
Ack. Added version checking. But please, notice that due to this switching logic, I preferred to use the `set_target_properties(INTERFACE_INCLUDE_DIRECTORIES)` so as to avoid repeating the full `install()` command in the two branches of the `if` statement.

ruslo commented:
Current version is not working on Windows with Visual Studio 2017. I've noticed that both `GTEST_CREATE_SHARED_LIBRARY` and `GTEST_LINKED_AS_SHARED_LIBRARY` used on compile step.

Generate:
```
> cmake -H. -B_builds -G "Visual Studio 15 2017" -DBUILD_SHARED_LIBS=ON -Dgtest_force_shared_crt=ON
```

Build:
```
> cmake --build _builds --config Release
```

Error message:
```
ClCompile:
  ...\CL.exe ... /D GTEST_CREATE_SHARED_LIBRARY=1 /D GTEST_LINKED_AS_SHARED_LIBRARY=1 ...\gtest_main.cc
...\googletest\googletest\src\gtest_main.cc(34): error C2491: 'main': definition of dllimport function not allowed
```

As I said before `GenerateExportHeader` helper module can resolve it for you and hide all implementation details of using `__declspec(dllimport)` and `__declspec(dllexport)`.

stefanosoffia commented:
Ahh! Well spotted. Fixed now. There was a missing `$<INSTALL_INTERFACE>`. Please, notice that `GenerateExportHeader` is not available with other build systems which should be supported by this project, i.e. bazel and GNU autoconf. So, I would fix this way for this PR.

I am a bit surprised that this issue was not caught by the AppVeyor CI, but now I see that the regression only covers static libraries on Windows. Shall/Can we do something about this? i.e. are we allowed to add new tests to the regression?

OT: @ruslo I'm not particularly satisfied by the fact that we have two separate packages for GTest and GMock. What do you think about having a single CMake package for all libraries? This would also simplify the handling of dependencies of gmock on gtest on Linux.

ruslo commented:
> Fixed now

Example is not working with your patches:
* https://github.com/ruslo/hunter/blob/3c5a5d8a8bd3c4e0a27b1f5175aca34d3796b6a7/examples/GTest/foo.cpp

Visual Studio 15 2017 with DLL.

I see warnings:
```
...\include\gmock\gmock-spec-builders.h(866): warning C4251:
  'testing::internal::ExpectationBase::untyped_actions_': class 'std::vector<const void *,std::allocator<_Ty>>'
  needs to have dll-interface to be used by clients of class 'testing::internal::ExpectationBase'
```

The worst part is that all my tests silently ignored:
```
Running main() from gtest_main.cc
[==========] Running 0 tests from 0 test cases.
[==========] 0 tests from 0 test cases ran. (0 ms total)
[  PASSED  ] 0 tests.
```

> Please, notice that GenerateExportHeader is not available with other build systems which should be supported by this project, i.e. bazel and GNU autoconf

Since we are discussing CMake here I'm talking about right way to do it in CMake. I don't know how to solve it in other build systems.

> What do you think about having a single CMake package for all libraries?

From my understanding one project - one `*Config.cmake` file. So current repository should produce targets like `googletest::gtest_main` and `googletest::gmock_main`. But as a side note I don't understand why these two projects was merged in the first place. As far as I remember there was two different repositories before.

stefanosoffia commented:
@ruslo, thanks for the pointer to your example. To build it I used the following `CMakeLists.txt`:
```
cmake_minimum_required(VERSION 3.0)
project(download-gtest)
enable_testing()

set(gtest_custom_install_dir "" CACHE PATH "")
list(APPEND CMAKE_PREFIX_PATH "${gtest_custom_install_dir}")
find_package(GMock CONFIG REQUIRED)
add_executable(foo foo.cpp)
target_link_libraries(foo PUBLIC GMock::gmock_main)

add_test(NAME foo_test COMMAND foo)
set_tests_properties(foo_test PROPERTIES ENVIRONMENT
  "PATH=$<TARGET_FILE_DIR:GMock::gmock_main>")
```
It compiles _fine_, i.e. it has warnings, and the test runs succesfully:
```
D:\hunter\examples\GTest\build>ctest -V
UpdateCTestConfiguration  from :D:/hunter/examples/GTest/build/DartConfiguration
.tcl
UpdateCTestConfiguration  from :D:/hunter/examples/GTest/build/DartConfiguration
.tcl
Test project D:/hunter/examples/GTest/build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 1
    Start 1: foo_test

1: Test command: D:\hunter\examples\GTest\build\foo.exe
1: Environment variables:
1:  PATH=D:/googletest/install/bin
1: Test timeout computed to be: 9.99988e+06
1: Running main() from gmock_main.cc
1: [==========] Running 2 tests from 1 test case.
1: [----------] Global test environment set-up.
1: [----------] 2 tests from Foo
1: [ RUN      ] Foo.Boo
1: [       OK ] Foo.Boo (0 ms)
1: [ RUN      ] Foo.Mock
1: [       OK ] Foo.Mock (0 ms)
1: [----------] 2 tests from Foo (0 ms total)
1:
1: [----------] Global test environment tear-down
1: [==========] 2 tests from 1 test case ran. (0 ms total)
1: [  PASSED  ] 2 tests.
1/1 Test #1: foo_test .........................   Passed    0.12 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   0.13 sec
```

Some thoughts:
* Notice the `set_tests_properties()` to add the `gmock` shared library folder to the path. 
* AFAIK, due to the current build strategy for Windows of `google/googletest`, if you link `gmock` then you shall not link `gtest`.  
* About the warnings, I reviewed them and I think that these cannot be easily solved without major changes to the sources. Look for example at this one:
```
D:\googletest\install\include\gtest\internal\gtest-internal.h|147| warning C4275: non dll-interface class 'std::runtime_error' used as base for dll-interface class 'testing::internal::GoogleTestFailureException'
```
This belongs to the C++ standard library. See [this thread](https://stackoverflow.com/questions/2132747/warning-c4251-when-building-a-dll-that-exports-a-class-containing-an-atlcstrin). Still I'm not sure about the solution.

I'm going to modify this PR so as to generate a single CMake package.

ruslo commented:
> if you link gmock then you shall not link gtest

Example use both `#include <gtest/gtest.h>` and `#include <gmock/gmock.h>` so it's okay to link both GTest and GMock libraries. If adding GTest libraries change behavior of my tests then it means GTest sources has defect in CMake configuration.

> Still I'm not sure about the solution

I haven't tried but I'm pretty sure that using `GenerateExportHeader` will do the trick. There may be other solutions, I'm not saying that it is the only one, but it is the right and easy one for CMake.

stefanosoffia commented:
@ruslo, let me understand because now I'm not following you. From what I see, on Windows, `gmock` sources include also `gtest` sources. Let's assume that right now we don't want/can't change this. Therefore you have to either link `gmock` or `gtest`, but not both. I don't think this is a defect of the CMake target export logic. Even under Linux, with this specific example you _can_ link `gmock` only because the `gtest` is then implicitly inherited.

About the second part, to verity this I'm going to create a self contained example based on `GenerateExportHeader` and report about it here.

ruslo commented:
> Therefore you have to either link gmock or gtest, but not both.

If you think that it is an error, then create such CMake code that will report an error.

> I don't think this is a defect of the CMake target export logic

Yes it is.

> About the second part, to verity this I'm going to create a self contained example based on GenerateExportHeader and report about it here

👍

stefanosoffia commented:
@ruslo, thank you again for your time.
About the export header, this is the self-contained example I come up with:
```
cmake_minimum_required(VERSION 3.0)
project(generate_export_example)
enable_testing()
include(GenerateExportHeader)

add_library(mylib SHARED mylib.cpp)
generate_export_header(mylib)
target_include_directories(mylib
  PUBLIC . ${CMAKE_CURRENT_BINARY_DIR})

add_executable(myexe myexe.cpp)
target_link_libraries(myexe PUBLIC mylib)
add_test(NAME mytest COMMAND myexe)
```
This is `mylib.h`:
```
#pragma once
#include <mylib_export.h>
#include <string>

class MYLIB_EXPORT SharedUtility {
public:
  void do_something();

private:
  std::string state;
};
```
This is `mylib.cpp`:
```
#include <mylib.h>
#include <iostream>

void SharedUtility::do_something() {
  state.append("+");
  std::cout << "SharedUtility" << state << "\n";
}
```
This is `myexe.cpp`:
```
#include <mylib.h>

int main() {
  SharedUtility u;
  u.do_something();
  u.do_something();
  u.do_something();
  return 0;
}
```
This is the warning I get when I build it on Windows:
```
D:\tmp\cmake_export_example\mylib.h(10): warning C4251: 'SharedUtility::state':
class 'std::basic_string<char,std::char_traits<char>,std::allocator<char>>' need
s to have dll-interface to be used by clients of class 'SharedUtility'
C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\VC\Tools\MSVC\14.
12.25827\include\xstring(4054): note: see declaration of 'std::basic_string<char
,std::char_traits<char>,std::allocator<char>>'
```
I think this reproduces fairly what happens in GTest, and it uses `GenerateExportHeader`. I observed that the warning can be solved just by moving the `MYLIB_EXPORT` from the `class` to the `do_something()` mehtod. However, this is independent of the `GenerateExportHeader`, i.e. the same effect can be obtained by moving `_GTEST_API` from the class to the methods. What do you think?

About the task of creating a CMake code that will report an error in case both `gtest` and `gmock` are linked, it would be great, but I have to admit that I have no idea how to realize it. Can this be obtained with CMake only? or shall we do something in the library source?

ruslo commented:
> I think this reproduces fairly what happens in GTest, and it uses GenerateExportHeader.

Okay, it seems that we should blame STL stuff for this warning. I guess that's why I haven't seen such warning in my example. But now I don't know how to explain the bad behavior :)

> Can this be obtained with CMake only? or shall we do something in the library source?

Now I don't know what is the problem exactly, so hard to say what is the solution. From user perspective the problem here is that you can't just say "please don't link GTest::gtest when you already linker GMock::gmock" (is it documented by the way?).

As an example:
```cmake
option(WITH_EXTRA_TESTS "Enable extra tests" OFF)

set(sources basic_tests.cpp)
if(WITH_EXTRA_TESTS)
  list(APPEND sources extra_tests.cpp)
endif()

add_executable(boo ${sources})

find_package(GTest CONFIG REQUIRED)
target_link_libraries(boo PUBLIC GTest::gtest_main)

if(WITH_EXTRA_TESTS)
  find_package(GMock CONFIG REQUIRED)
  target_link_libraries(boo PUBLIC GMock::gmock)
endif()
```

For me look like a straightforward approach to extend your set of tests, results:
* static build
  * `WITH_EXTRA_TESTS=OFF` works as expected
  * `WITH_EXTRA_TESTS=ON` words as expected
* shared build
  * `WITH_EXTRA_TESTS=OFF` works as expected
  * `WITH_EXTRA_TESTS=ON` **no tests run**, failed tests **ignored**, success status was returned

stefanosoffia commented:
I don't know either what is the problem exactly. I am currently unable to reproduce it with a self-contained sample. But I noticed that at least on my machine (VS2017) it is enough to swap the order of the two `target_link_libraries()` to make the generated exe work. My impression is that the _original sin_ is to try to linking together two shared that are exporting the same symbols. I feel it as something equivalent to the notion of _undefined behavior_ of C++. Also [the doc here](https://github.com/google/googletest/blob/master/googlemock/README.md#building-google-mock) seems to suggest that we should link just one library.

ruslo commented:
> I am currently unable to reproduce it with a self-contained sample

Take a look at this example:
* https://github.com/forexample/gtest-conflict

Builds on AppVeyor Windows machines:
* https://ci.appveyor.com/project/ruslo/gtest-conflict/build/1.0.4

As you can see all combinations work fine except `WITH_EXTRA_TESTS=ON` + `BUILD_SHARED_LIBS=ON`:
* https://ci.appveyor.com/project/ruslo/gtest-conflict/build/1.0.4/job/2dhswyphd43xoqfh

```
1: [==========] Running 0 tests from 0 test cases.
1: [==========] 0 tests from 0 test cases ran. (0 ms total)
1: [  PASSED  ] 0 tests.
```

> My impression is that the original sin is to try to linking together two shared libraries that are exporting the same symbols

I guess I missed that. Yes, that's the problem. 

Currently target-sources relations:

* gtest_main.dll <- gtest_main.cc
* gtest.dll <- gtest-all.cc
* gmock_main.dll <- gtest-all.cc gmock-all.cc gmock_main.cc
* gmock.dll <- gtest-all.cc gmock-all.cc

From my opinion it should be:

* gtest_main.dll <- gtest_main.cc
* gtest.dll <- gtest-all.cc
* gmock_main.dll <- gmock_main.cc
* gmock.dll <- gmock-all.cc

And dependencies between libraries should be expressed by `target_link_libraries`.

Note that current version is at least not optimal, e.g. we have to compile gtest-all.cc 3 times with the same compiler flags. Actually I have reported this issue before: https://github.com/google/googletest/issues/686

> Also the doc here seems to suggest that we should link just one library

It just saying that if you want to use GMock you can change gtest_main to gmock_main. It's not saying that linking gmock to gtest leads to incorrect behavior.

stefanosoffia commented:
@ruslo, as anticipated I updated this patch so as to export just one CMake package.

Now, I was going through the past discussion trying to create a proposal for a list of activities. Right now, this is what comes to my mind:
1. Optional but AFAIK advisable: Update the `cmake_minimum_required()` and cleanup the `CMakeLists.txt` from all the `VERSION_LESS` switches. A bump to `3.1` should allow to remove all such switches.
2. Improve the handling of the dll exports for Windows. I would say that the end goal would be to have the same specification for both Linux and Windows. But for this step 2 I would focus on just replacing the current hand crafted macros with something better. We have a constraint here: The `google/googletest` shall be able to build with other build systems, e.g. autotools and bazel. The `GenerateExportHeader` is solid. We should check if and how the generated `_export.h` files can be versioned and reused by other build systems to be supported. Just for completeness, this talk [CppCon 2017: James McNellis “Everything You Ever Wanted to Know about DLLs”](https://www.youtube.com/watch?v=JPQWQfDhICA&t=3290s) also brings to the attention on the use of external `.defs` files. I'm not convinced. Also, CMake supports the [CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS](`https://cmake.org/cmake/help/latest/variable/CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS.html`), which by the way seems a bit gross and not sure how to support other build systems with this one.
3. Eventually, reorganize the libraries structure and dependencies as you summarized.

I was thinking of one PR for each point. Once we find a sort of an agreement on this _step 0_ we should look for someone in the team. WDYT?

ruslo commented:
> A bump to 3.1 should allow to remove all such switches

👍 

> We should check if and how the generated _export.h files can be versioned and reused by other the build systems to be supported

I think it depends on 3: one of the things that `GenerateExportHeader` will bring is that you should use 4 different macros `_GTEST_API`, `_GTEST_MAIN_API`, `_GMOCK_API` and `_GMOCK_MAIN_API` instead of one `_GTEST_API`.

> Eventually, reorganize the libraries structure and dependencies as you summarized

👍

ruslo commented on googletest/CMakeLists.txt:
```diff
@@ -72,6 +72,25 @@ include(cmake/internal_utils.cmake)
+if (INSTALL_GTEST)
+  include(CMakePackageConfigHelpers)
+  set(cmake_package_name GTest)
+  set(targets_export_name ${cmake_package_name}Targets CACHE INTERNAL "")
```
I think it make sense just to move this code to top-level CMakeLists.txt, in this case there is no need to use `CACHE INTERNAL` variables. Also this file is optional, we will not reach this code if `BUILD_GTEST` is `OFF`.

stefanosoffia replied:
@ruslo, what you suggest has been my first attempt because it is logical. Though, I had to revert it to the current code because:
1. `googlemock/CMakeLists.txt` actually `add_subdirectory()` the `gtest` project. Therefore the above code fragment is always included, even when `BUILD_GTEST=OFF`.
2. To be able to inject the `GTEST_HAS_PTHREAD` into the generated `Config.cmake` file, this code shall be included _after_ the call to `config_compiler_and_linker()`, which is currently in `googletest/CMakeLists.txt`. At the same time it shall be _before_ the calls to our `install_project()`. Putting it in the top-level `CMakeLists.txt` does not satisfy either one or the other requirement.

WDYT?

stefanosoffia replied:
Afterthought: AFAIK `BUILD_GTEST` is vestigial from the time they were two separate projects, i.e. with different lifecycles. IMHO, the case `BUILD_GTEST=OFF AND BUILD_GMOCK=OFF` is not very interesting and could be simplified away by dropping `BUILD_GTEST` completely. This is an item for the list of next steps.

ruslo replied:
> not very interesting and could be simplified away by dropping BUILD_GTEST completely

👍 

> googlemock/CMakeLists.txt actually add_subdirectory() the gtest project

> this code shall be included after the call to config_compiler_and_linker()

I think this code should be refactored.

stefanosoffia replied:
I removed `BUILD_GTEST`. Further reorganization is possible but the side effects are not trivial to follow. Let's discuss them later.

ruslo replied:
> Let's discuss them later

From my opinion they should be fixed **before** this pull request proceed.

stefanosoffia replied:
@ruslo, I did some changes in this direction, including probably something too much. But let's start the discussion. I split the changes in more commits trying to keep functionally related stuff together. This grouping is subject to discussion too.

Side note: I notice that previously, two of the python tests were not detected by CMake, probably due to the strange inclusion order of the projects. These are `gmock_leak_test` and `gmock_output_test`.

ruslo replied:
> I did some changes in this direction, including probably something too much

Yes, I don't think it is a good idea to mix everything into one big change. For review and for testing reasons it will be better to create several pull requests, e.g. "Set CMake 3.1 as required minimum".

gennadiycivil commented:
I am not sure that googltest can make CMake 3.1 required. This warrants wide discussion between users which may or may not be practical.

stefanosoffia commented:
@gennadiycivil, let me refocus this conversation and PR on the first commit only, which enables GoogleTest to be included as CMake package by users. I now discarded the commits about the CMake minimum version bump and the CMake projects cleanup which were the result of the conversation.

As suggested by @ruslo I'll propose a separate PR for the CMake cleanup.

ruslo commented:
> I'll propose a separate PR for the CMake cleanup

👍

ccvca commented:
If the CMake Requirement should not be >=3.1, maybe its possible to enable/disable some parts of install depending on the CMake version (https://stackoverflow.com/questions/43697154/detect-current-cmake-version-using-cmake)?

gennadiycivil commented:
There appear to be CLA problems. Can you confirm that "all authors are ok with their commits being contributed to this project. Please have them confirm that here in the pull request."

gennadiycivil commented:
Also please resolve conflicts

gennadiycivil commented:
@stefanosoffia @ruslo @sebastianlipponer @ccvca . This looks OK to me, any objections to merge it?

stefanosoffia commented:
Updated to integrate the recently merged #1682.

dakotahawkins commented:
@stefanosoffia @gennadiycivil This seems to revert a LOT of (all of?) the work from #1682, which seems like a step backwards.

I don't see any discussion around that, so what happened? I'd rather help build on #1682 than throw it away :(

stefanosoffia commented:
@dakotahawkins sorry, I guess then I probably missed part of the intent of PR #1628.

When I merged PR #1628 I did some changes to #1338 to account for some of your ideas, e.g. to have `INTERFACE_INCLUDE_DIRECTORIES` populated also when `googletest` is included by `add_subdirectory`.

But let's leverage this momentum! 

AFAIK the root cause for these sort of misunderstanding is that there are no tests that document the requirements by exercising the various `add_subdirectory`/`find_package` combinations. I do have two sample projects that perform this, though at the moment it is still manual.

On the PR side let's switch to #1628 for the detailed discussion.

stefanosoffia commented:
Let's proceed by steps. I just pushed two commits to address:

1. Names of the installed targets:
 * Be consistent with the widespread `FindGTest` module.
 * Let the same namespaced target names be available both when `googletest` is used by `find_package` and `add_subdirectory`.
2. Added cache variable to control the location of the exported targets config file. Though, I left the default value to be the same for Windows and Linux. Let me explain my concern with the distinct defaults for Windows and Linux. AFAIK the case of downloading the sources, compiling them and installing locally is more frequent than installing system wide. E.g. see CMake `ExternalProject_Add`. To ease this use case, I prefer to keep the default the same because it simplifies portability. Anyway, this is not an absolute, so please discuss this together and see if there are other reasons to keep the distinction.

stefanosoffia commented:
@ruslo, can you please point me to the review note you were mentioning?

dakotahawkins commented:
@stefanosoffia I think it probably makes more sense to keep them the same for Windows and Linux by default, I guess my preference on Windows might not be good enough to make it a default.

It _should_ be a cache variable though, so it can be customised and I can live with it :)

stefanosoffia commented:
@dakotahawkins back to your list, what is the next point you would like improve? Can you help me better undestand the points you listed?

dakotahawkins commented:
@stefanosoffia I've taken a stab at what I want in #1785, let's move the discussion there (and please review the changes)

FYI @gennadiycivil
